### PR TITLE
ACCUMULO-3626: Reset security should warn and require instance secret

### PR DESCRIPTION
**Pull request by milleruntime**
Created a prompt when running "accumulo init --reset-security" to warn user and require
entering the instance secret.  Added a "-f,--force" option for scripts to bypass the prompt.

Also fixed message in Initialize about running ChangeSecret tool.

Ticket fix version says 2.0.0 but I think 1.8 and other versions would benefit as well.

### CODE REVIEW

File: server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java [hunk 1]
```diff
@@ -238,7 +238,7 @@ static boolean checkInit(Configuration conf, VolumeManager fs, SiteConfiguration
       c.println();
       c.println();
       c.println("You can change the instance secret in accumulo by using:");
-      c.println("   bin/accumulo " + org.apache.accumulo.server.util.ChangeSecret.class.getName() + " oldPassword newPassword.");
+      c.println("   bin/accumulo " + org.apache.accumulo.server.util.ChangeSecret.class.getName());
```
Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: This change seems unrelated to your description.

File: server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java [hunk 2]
```diff
@@ -762,6 +764,17 @@ public void execute(final String[] args) {
       if (opts.resetSecurity) {
         AccumuloServerContext context = new AccumuloServerContext(new ServerConfigurationFactory(HdfsZooInstance.getInstance()));
         if (isInitialized(fs)) {
+          if (!opts.forceResetSecurity) {
+            String secret = acuConf.get(Property.INSTANCE_SECRET);
+            ConsoleReader c = getConsoleReader();
+            String userEnteredSecret = c.readLine("WARNING: This will remove all users from Accumulo! If you wish to proceed enter the instance secret: ", '*');
```
Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: > This will remove all users from Accumulo

Really, this is removing all ZooKeeper data. The "user database" just happens to be in ZooKeeper. Maybe there's a better way to word this?

Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: > If you wish to proceed enter the instance secret

Do we need to prompt for instance.secret instead of just a 'Y/N'? It's not apparent if this is buying us anything. If someone can successfully re-init (permission wise), they have the ability to get this value.

File: server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java [hunk 3]
```diff
@@ -762,6 +764,17 @@ public void execute(final String[] args) {
...
+            if (userEnteredSecret != null && !secret.equals(userEnteredSecret)) {
+              log.info("Aborted reset security: Instance secret did not match what is in accumulo-site.xml");
```
Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: s/log.info/log.error/

File: server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java [hunk 4]
```diff
@@ -762,6 +764,17 @@ public void execute(final String[] args) {
...
+              return;
+            }
+            log.info("Beginning user initiated reset security on accumulo.");
```
Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: Maybe implied?

**Comment by joshelser** [comment]
> Ticket fix version says 2.0.0 but I think 1.8 and other versions would benefit as well.

I am hesitant to include this on patch-releases as this is a very user-facing change. This could break automatic provisioning tools (e.g. Ambari, Cloudera Manager, etc).

File: server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java (hunk 1 above)
Comment (milleruntime) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: Correct. I noticed the message was wrong and fixed it while editing the file.

File: server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java (hunk 2 above)
Comment (milleruntime) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: I am open to suggestions for the wording of the message, I was going by what was requested in the ticket.  I believe having the user enter something other than Y/N was to require additional thought as to whether they really want perform this action.

Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: The prompt for the secret is similar to how GitHub asks you to type the name of the repo, to be sure you really want to delete it. It's not a security check. The user running this command already needs access to the server configuration file containing the secret. It's just a sanity check. Instead of the secret, it could prompt for the instance name or ID.

File: server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java (hunk 4 above)
Comment (milleruntime) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: I added this log message since it is not clear what is going on in the output...
2016-09-14 12:35:03,251 [conf.AccumuloConfiguration] INFO : Loaded class : org.apache.accumulo.server.security.handler.ZKAuthorizor
2016-09-14 12:35:03,252 [conf.AccumuloConfiguration] INFO : Loaded class : org.apache.accumulo.server.security.handler.ZKAuthenticator
2016-09-14 12:35:03,253 [conf.AccumuloConfiguration] INFO : Loaded class : org.apache.accumulo.server.security.handler.ZKPermHandler
2016-09-14 12:35:03,508 [handler.ZKAuthenticator] INFO : Removed /accumulo/5a96e1b5-452b-4c35-ada7-cfa807cbd994/users/ from zookeeper

**Comment by ctubbsii** [comment]
Because this is a pretty dangerous command to run without user interaction, I'd be fine adding it in 1.7 and 1.8 to add a bit of safety, so long as we clearly document in the release notes that automated systems will need to use the force flag.

File: server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java (hunk 4 above)
Comment (milleruntime) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: Should be outside the if though... I can remove or move to the beginning of if (opts.resetSecurity) block?

File: server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java (hunk 2 above)
Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: > it could prompt for the instance name or ID

I like "instance name" instead of the secret. I do not believe the secret is self-identifying WRT the Github analogy you make.

File: server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java (hunk 4 above)
Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: Maybe instead of "Beginning user initiated reset security on accumulo" just "Resetting security on Accumulo".

**Comment by joshelser** [comment]
> Because this is a pretty dangerous command to run without user interaction

I would also say that I think the risk is rather low to "stumble" onto this. I would guess the risk of an user/admin accidentally running this command is low. I think our docs do a good job already alerting that this is a hammer, not a scalpel (although this is off of memory)

File: server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java (hunk 2 above)
Comment (busbey) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: yeah, I like typing the instance name as well. it lets us echo the response as they type, and it clearly identifies what is is they're saying to delete (not that anyone would reuse instance secrets ;)).

**Comment by busbey** [comment]
I don't think this would break Cloudera Manager specifically, since this is the kind of thing we'd document in Knowledge Base articles for support / customers rather than try to automate (as mentioned, it's dangerous and a blunt tool).

But I'd also rather see this in a minor or major release and not a maintenance release since it changes how an operator facing tool works. I've already documented / explained / conditioned folks to expect this tool to be useful in a set of circumstances and to act a certain way (again, since it's dangerous we train folks to use it by saying what to expect and not expect as they use it). I won't be able to grab maintenance releases in a straight-forward way if I have to update internal training things about this tool to specify different expectations for e.g. 1.7.2 vs 1.7.3.

File: server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java (hunk 2 above)
Comment (milleruntime) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: Fixed in d2b1ddc3b42dde1550000244c8bdc67c98d31c07

File: server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java (hunk 3 above)
Comment (milleruntime) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: Fixed in d2b1ddc3b42dde1550000244c8bdc67c98d31c07

File: server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java (hunk 4 above)
Comment (milleruntime) [inline server/base/src/main/java/org/apache/accumulo/server/init/Initialize.java]: Fixed in d2b1ddc3b42dde1550000244c8bdc67c98d31c07

**Comment by milleruntime** [comment]
I changed this to merge into master.  If there aren't any reservations to have this released in 2.0.0, then can it be merged?

**Comment by joshelser** [comment]
>  If there aren't any reservations to have this released in 2.0.0, then can it be merged?

:+1: from me.

**Comment by milleruntime** [comment]
@busbey You OK with this change going out in 2.0.0?

**Comment by busbey** [comment]
+1 for 2.0.0
